src/kesslergame/ship.py

```diff
@@ -196,45 +239,146 @@ def update(self, delta_time: float = 1 / 30) -> tuple[Optional[Bullet], Optional
             warnings.warn('Ship ' + str(self.id) + ' thrust command outside of allowable range', RuntimeWarning)
 
-        # Apply thrust
-        self.speed += self.thrust * delta_time
-
-        # Bounds check the speed
-        if self.speed > self.max_speed:
-            self.speed = self.max_speed
-        elif self.speed < -self.max_speed:
-            self.speed = -self.max_speed
-
         # Bounds check the turn rate
         if self.turn_rate < self.turn_rate_range[0] or self.turn_rate > self.turn_rate_range[1]:
             self.turn_rate = min(max(self.turn_rate_range[0], self.turn_rate), self.turn_rate_range[1])
             warnings.warn('Ship ' + str(self.id) + ' turn rate command outside of allowable range', RuntimeWarning)
 
+        # Store speed and heading BEFORE acceleration/thrust for integration
+        initial_speed = self.speed
+        theta0 = math.radians(self.heading)  # convert to radians
+
+        is_moving: bool = abs(initial_speed) > 1e-12
+
+        # Get direction of motion for drag
+        if abs(initial_speed) > 1e-12:
+            motion_sign = math.copysign(1.0, initial_speed)
+        else:
+            motion_sign = math.copysign(1.0, self.thrust) if abs(self.thrust) > 1e-12 else 0.0
+
+        drag_acc = -self.drag * motion_sign
+
+        # Combine thrust and drag into one net acceleration
+        net_acc = self.thrust + drag_acc  # m/s²
+        
+        # Analytic position integration, framerate independent
+        # The shape that the ship traces out with a constant turn rate and thrust over the previous frame is a type of spiral
+        # This spiral can be analytically integrated! Yay!
+        # The Sympy code to set up dynamics and integrate is as follows:
+
+        # from sympy import *
+        # x0, y0, v0, theta0, omega, delta_x, delta_y, t, delta_t, a = symbols('x0 y0 v0 theta0 omega delta_x delta_y t delta_t a')
+        # v_t = v0 + a * t
+        # theta_t = theta0 + omega * t
+        # delta_x_expr = integrate(v_t * cos(theta_t), (t, 0, delta_t))
+        # delta_y_expr = integrate(v_t * sin(theta_t), (t, 0, delta_t))
+
+        # And notice how after integrating, we can simplify things by defining: theta1 = theta0 + omega * delta_t
+
+        x0: float = self.x
+        y0: float = self.y
+        omega = math.radians(self.turn_rate)
+
+        def analytic_integration(vi: float, a: float, theta0: float, omega: float, t: float) -> tuple[float, float]:
+            """
+            Returns (dx, dy) using either analytic or Taylor expansion for small omega.
+            Args:
+                vi: initial speed
+                a: acceleration
+                th0: initial heading (radians)
+                omega: turn rate (rad/sec)
+                t: duration (seconds)
+            """
+            if abs(t) < 1e-12:
+                return 0.0, 0.0
+            if abs(omega) < 1e-4:
+                # Taylor expansion with Horner's method
```

Comment:
K this is totally wrong. I'll fix this once I do the math, but it's actually pretty tricky